File: python/day23.py
"""Day 23: Experimental Emergency Teleportation -- AoC 2018.

Algorithm reference in Python.  The shipping solution lives in
src/Day23.hs; this file is here so the *algorithm* reads without
Haskell's Set-as-priority-queue plumbing in the way.

The input is a swarm of nanobots, each an (x, y, z) position and a
signal radius r.  A bot is "in range" of any integer point within
Manhattan distance r of it -- so each bot's coverage is an octahedron
(a taxicab ball).

  * Part 1: the bot with the largest radius; how many bots (itself
    included) lie within its range.  One linear scan.

  * Part 2: the integer point in range of the *most* bots; among ties,
    the one closest to the origin; report that Manhattan distance.

Part 2 is the calendar's hardest search -- the answer can be anywhere
in a ~10^8-wide cube, far too large to scan.  The trick is branch and
bound over an *octree*:

  * Bound every bot inside one big power-of-two cube.
  * For a cube, count how many bots' ranges *touch* it.  No point
    inside can be seen by more bots than that, so the count is an
    admissible upper bound.
  * Keep cubes in a priority queue ordered (most bots, then closest to
    origin, then smallest).  Pop the best, split it into 8 octants,
    push them back.  Because the bound only shrinks and the origin
    distance only grows as cubes get smaller, the first 1x1x1 cube
    pulled from the queue is provably the optimal point.

Run from the repo root:

    python python/day23.py

Reads inputs/day23.txt and prints both parts.
"""
import heapq
import re
from pathlib import Path

INPUT = Path(__file__).parent.parent / "inputs" / "day23.txt"

NUM = re.compile(r"-?\d+")


def parse(text):
    """Each line -> ((x, y, z), r).  Pull the four integers out."""
    bots = []
    for line in text.strip().splitlines():
        x, y, z, r = map(int, NUM.findall(line))
        bots.append(((x, y, z), r))
    return bots


def manhattan(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1]) + abs(a[2] - b[2])


# ---------------------------------------------------------------------------
# Part 1
# ---------------------------------------------------------------------------

def part1(bots):
    pos, r = max(bots, key=lambda b: b[1])          # strongest = largest radius
    return sum(1 for p, _ in bots if manhattan(pos, p) <= r)


# ---------------------------------------------------------------------------
# Part 2: branch and bound over an octree
# ---------------------------------------------------------------------------

def dist_point_box(p, lo, size):
    """Manhattan distance from point p to the nearest cell of the cube
    [lo .. lo+size-1]^3.  Per axis: how far p lies outside the range."""
    total = 0
    for v, l in zip(p, lo):
        h = l + size - 1
        if v < l:
            total += l - v
        elif v > h:
            total += v - h
    return total


def dist_box_origin(lo, size):
    """Manhattan distance from the origin to the nearest cell of the cube."""
    total = 0
    for l in lo:
        h = l + size - 1
        if l > 0:
            total += l
        elif h < 0:
            total += -h
    return total


def part2(bots):
    xs = [p[0] for p, _ in bots]
    ys = [p[1] for p, _ in bots]
    zs = [p[2] for p, _ in bots]
    lo = (min(xs), min(ys), min(zs))
    extent = max(max(xs) - lo[0], max(ys) - lo[1], max(zs) - lo[2])
    size = 1
    while size <= extent:
        size *= 2                                    # smallest power of two > extent

    def count(lo, size):
        return sum(1 for p, r in bots if dist_point_box(p, lo, size) <= r)

    # Heap key: (-count, dist_to_origin, size, lo).  Python's tuple order
    # makes this "most bots, then closest, then smallest cube" first.
    heap = [(-count(lo, size), dist_box_origin(lo, size), size, lo)]
    while heap:
        neg_c, d, size, lo = heapq.heappop(heap)
        if size == 1:
            return d                                 # first 1x1x1 popped is optimal
        h = size // 2
        for dx in (0, h):
            for dy in (0, h):
                for dz in (0, h):
                    nlo = (lo[0] + dx, lo[1] + dy, lo[2] + dz)
                    heapq.heappush(
                        heap,
                        (-count(nlo, h), dist_box_origin(nlo, h), h, nlo),
                    )
    raise RuntimeError("queue emptied without reaching a 1x1x1 cube")


def main():
    bots = parse(INPUT.read_text())
    print("part 1:", part1(bots))
    print("part 2:", part2(bots))


if __name__ == "__main__":
    main()

File: src/Day23.hs
{-# LANGUAGE BangPatterns #-}

-- |
-- Module      : Day23
-- Description : Day 23 -- Experimental Emergency Teleportation
--
-- A swarm of /nanobots/, each with an integer position @(x,y,z)@ and a
-- /signal radius/ @r@.  A bot is /in range/ of any integer coordinate
-- within Manhattan distance @r@ of its position.
--
--   * Part 1: find the bot with the largest radius; count how many
--     bots (including itself) lie within that bot's range.
--
--   * Part 2: find the integer coordinate that is in range of the
--     /most/ bots; among all such coordinates, take the one closest to
--     the origin; report that Manhattan distance.
--
-- Concepts introduced this day:
--
--   * /Manhattan (taxicab) balls/ in 3-D.  A bot's coverage is an
--     octahedron @|x-px| + |y-py| + |z-pz| <= r@.  We never enumerate
--     points -- the whole puzzle is about reasoning over these balls.
--
--   * /Branch and bound over an octree/.  Part 2's search space is all
--     of 3-space, far too large to scan.  Instead we bound a cube,
--     compute an /upper bound/ on how many bots any point inside it
--     could see, and recursively split the most promising cube into 8
--     octants until a single cell survives.  The first 1x1x1 cube
--     pulled from the queue is provably optimal.
--
--   * 'Data.Set' as a priority queue with the search key chosen so the
--     /lexicographic/ 'Ord' on a 4-tuple encodes the whole policy:
--     most bots first, then closest to origin, then smallest cube.
--     Same Set-as-PQ pattern as Day 22's Dijkstra.
--
--   * /Point-to-box Manhattan distance/ via per-axis clamping: the
--     nearest cell of an axis-aligned box to a point costs
--     @max 0 (lo - v) (v - hi)@ per axis, summed.  This is the bound
--     that makes the branch-and-bound admissible.
module Day23
  ( -- * Types
    V3
  , Nanobot (..)
  , Puzzle

    -- * Public answers
  , parseInput
  , part1
  , part2
  , solve

    -- * Building blocks (exposed for tests)
  , manhattan
  , botsInRangeOf
  ) where

import           Control.DeepSeq (NFData (..))
import           Data.Char       (isDigit)
import           Data.List       (foldl', maximumBy)
import           Data.Ord        (comparing)
import qualified Data.Set        as Set

-- ---------------------------------------------------------------------------
-- Types
-- ---------------------------------------------------------------------------

-- | An integer point in 3-space.  A plain tuple gives us 'Eq'/'Ord'
-- (lexicographic) and a ready-made 'NFData' for free.
type V3 = (Int, Int, Int)

-- | One nanobot: where it sits and how far it can signal.  Strict
-- fields (@!@) keep the parsed swarm fully evaluated -- there is no
-- benefit to deferring a handful of 'Int's, and it avoids thunks
-- piling up behind the millions of distance checks Part 2 performs.
data Nanobot = Nanobot
  { botPos :: !V3
  , botR   :: !Int
  } deriving (Eq, Show)

-- | The parsed swarm.
type Puzzle = [Nanobot]

-- | Manual 'NFData' so criterion can @deepseq@ a parsed swarm.  The
-- tuple and 'Int' both already have instances; we just chain them.
instance NFData Nanobot where
  rnf (Nanobot p r) = rnf p `seq` rnf r

-- ---------------------------------------------------------------------------
-- Parsing
-- ---------------------------------------------------------------------------

-- | One 'Nanobot' per line, e.g. @pos=\<1,2,3\>, r=4@ (coordinates may
-- be negative).  Rather than thread a real parser through the angle
-- brackets and commas, we blank out every character that is not a
-- digit or a minus sign, then 'words' + 'read' the four integers that
-- remain.  Simple and total for this input shape.
parseInput :: String -> Puzzle
parseInput = map parseLine . lines

-- | Parse a single line into a 'Nanobot' by pulling out its integers.
parseLine :: String -> Nanobot
parseLine s = case ints s of
  [x, y, z, r] -> Nanobot (x, y, z) r
  _            -> error ("Day23.parseLine: cannot parse " ++ show s)

-- | Every signed integer appearing in a string, left to right.  We map
-- each character to itself if it is a digit or @\'-\'@, and to a space
-- otherwise, so @'words'@ then splits the integers apart cleanly.
ints :: String -> [Int]
ints = map read . words . map keep
 where
  keep c
    | isDigit c || c == '-' = c
    | otherwise             = ' '

-- ---------------------------------------------------------------------------
-- Geometry
-- ---------------------------------------------------------------------------

-- | Manhattan (taxicab) distance between two points.
manhattan :: V3 -> V3 -> Int
manhattan (x1, y1, z1) (x2, y2, z2) =
  abs (x1 - x2) + abs (y1 - y2) + abs (z1 - z2)

-- | How many bots have @target@ within their range (Part 1's counting
-- step, reused for clarity).
botsInRangeOf :: V3 -> Int -> Puzzle -> Int
botsInRangeOf centre r =
  length . filter (\b -> manhattan centre (botPos b) <= r)

-- ---------------------------------------------------------------------------
-- Part 1
-- ---------------------------------------------------------------------------

-- | Find the strongest bot (largest radius) and count how many bots
-- fall inside its signal range, itself included.
part1 :: Puzzle -> Int
part1 bots =
  let strongest = maximumBy (comparing botR) bots
  in  botsInRangeOf (botPos strongest) (botR strongest) bots

-- ---------------------------------------------------------------------------
-- Part 2: branch and bound over an octree
-- ---------------------------------------------------------------------------

-- | An axis-aligned cube: its minimum corner and side length.  The
-- side is always a power of two, so halving stays integral all the way
-- down to a single cell.  A cube of side @s@ at corner @(lx,ly,lz)@
-- covers the cells @[lx .. lx+s-1]@ in each axis.
data Box = Box !V3 !Int

-- | Manhattan distance from a point to the /nearest cell/ of a box.
-- Per axis the cost is how far @v@ lies outside @[lo,hi]@ (zero if
-- inside).  Summed over axes this is the closest a bot at @p@ could be
-- to any point of the box -- so @dist <= r@ means the bot's range
-- touches the box, i.e. it /could/ cover some point inside.
distPointBox :: V3 -> Box -> Int
distPointBox (px, py, pz) (Box (lx, ly, lz) s) =
  axis px lx (lx + s - 1) + axis py ly (ly + s - 1) + axis pz lz (lz + s - 1)
 where
  axis v lo hi
    | v < lo    = lo - v
    | v > hi    = v - hi
    | otherwise = 0

-- | Manhattan distance from the origin to the nearest cell of a box.
-- For a 1x1x1 box this is just that single cell's distance to origin,
-- which is the value Part 2 ultimately reports.
distBoxOrigin :: Box -> Int
distBoxOrigin (Box (lx, ly, lz) s) = axis lx (lx + s - 1)
                                   + axis ly (ly + s - 1)
                                   + axis lz (lz + s - 1)
 where
  axis lo hi
    | lo > 0    = lo
    | hi < 0    = negate hi
    | otherwise = 0

-- | Upper bound on coverage: how many bots' ranges intersect the box.
-- No point inside the box can be seen by more bots than this, and a
-- 1x1x1 box's bound is its exact coverage.
boxBound :: Puzzle -> Box -> Int
boxBound bots box = foldl' step 0 bots
 where
  step !acc (Nanobot p r)
    | distPointBox p box <= r = acc + 1
    | otherwise               = acc

-- | Split a cube into its eight half-size octants.
subdivide :: Box -> [Box]
subdivide (Box (lx, ly, lz) s) =
  [ Box (lx + dx, ly + dy, lz + dz) h
  | dx <- [0, h], dy <- [0, h], dz <- [0, h]
  ]
 where
  h = s `div` 2

-- | The smallest power-of-two cube that contains every bot position.
initialBox :: Puzzle -> Box
initialBox bots = Box (mnx, mny, mnz) size
 where
  xs            = [ x | Nanobot (x, _, _) _ <- bots ]
  ys            = [ y | Nanobot (_, y, _) _ <- bots ]
  zs            = [ z | Nanobot (_, _, z) _ <- bots ]
  (mnx, mny, mnz) = (minimum xs, minimum ys, minimum zs)
  extent        = maximum [ maximum xs - mnx
                          , maximum ys - mny
                          , maximum zs - mnz ]
  -- smallest power of two strictly greater than the extent, so the cube
  -- [mn .. mn+size-1] covers [mn .. mx] in every axis.
  size          = head (dropWhile (<= extent) (iterate (* 2) 1))

-- | The priority-queue key.  Lexicographic 'Ord' on this 4-tuple is
-- the entire search policy, smallest-first:
--
--   1. @negate count@  -- most bots first (largest coverage bound);
--   2. @distToOrigin@  -- ties broken toward the origin;
--   3. @size@          -- then prefer the smaller (more resolved) cube;
--   4. @corner@        -- final tiebreak, only to keep keys unique.
--
-- Storing @size@ and @corner@ lets us rebuild the 'Box' on pop without
-- a side table.
type Key = (Int, Int, Int, V3)

key :: Puzzle -> Box -> Key
key bots box@(Box lo s) =
  (negate (boxBound bots box), distBoxOrigin box, s, lo)

-- | Part 2.  Branch and bound: always expand the most promising cube.
-- Because the coverage bound only shrinks and the origin distance only
-- grows as cubes get smaller, the first 1x1x1 cube to reach the front
-- of the queue maximises coverage and, among those, minimises distance
-- to the origin -- exactly what the puzzle asks for.
part2 :: Puzzle -> Int
part2 bots = go (Set.singleton (key bots (initialBox bots)))
 where
  go pq = case Set.minView pq of
    Nothing -> error "Day23.part2: queue emptied without a 1x1x1 box"
    Just ((_, d, s, lo), rest)
      | s == 1    -> d
      | otherwise ->
          let children = subdivide (Box lo s)
              pq'      = foldl' (\q b -> Set.insert (key bots b) q) rest children
          in  go pq'

-- ---------------------------------------------------------------------------
-- Driver
-- ---------------------------------------------------------------------------

solve :: String -> IO ()
solve contents = do
  let puzzle = parseInput contents
  putStrLn ("  part 1: " ++ show (part1 puzzle))
  putStrLn ("  part 2: " ++ show (part2 puzzle))
